pkg/jobworker/worker.go

```diff
@@ -0,0 +1,208 @@
+	return StartWithController(&Cgroup{"/sys/fs/cgroup"}, opts, cmd, args...)
+}
+
+// StartWithController creates a job's cgroup, adds the resource controls from opts, creates a log file for the cgroup and
+// set's it to the exec.Cmd STDOUT and STDERR. Finally we Start the exec.Cmd and start a go routine to handle the blocking
+// call to Wait so that we can update the job's running flag.
+func StartWithController(con ResourceController, opts JobOpts, cmd string, args ...string) (j *Job, err error) {
+	// Create the job
+	j = NewJob(uuid.New().String(), exec.Command(cmd, args...), con)
+	// Create the cgroup and configure the controllers
+	if err = j.con.CreateGroup(j.ID); err != nil {
+		return nil, fmt.Errorf("failed to create cgroup: %w", err)
+	}
+	// Update cgroup controllers to add resource control to process
+	if err = j.con.AddResourceControl(j.ID, opts); err != nil {
+		return nil, fmt.Errorf("failed to add resource control: %w", err)
+	}
+	// Add job's process to cgroup
+	if err = j.con.AddProcess(j.ID, j.cmd); err != nil {
+		return nil, fmt.Errorf("failed to add PID to cgroup: %w", err)
+	}
+	defer syscall.Close(j.cmd.SysProcAttr.CgroupFD)
+	// Don't inherit environment from parent
+	j.cmd.Env = []string{}
+	// Pipe STDOUT and STDERR to a log file
+	f, err := os.OpenFile(logPath(j.ID), os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open job's log file: %w", err)
+	}
+	j.cmd.Stdout = f
+	j.cmd.Stderr = f
+	// Run the command as a given user as not to escalate privilege, since the executing user must also manage cgroups
+	if WORKER_UID != -1 && WORKER_GUID != -1 {
+		j.cmd.SysProcAttr.Credential = &syscall.Credential{Uid: uint32(WORKER_UID), Gid: uint32(WORKER_GUID)}
+	}
+	// Start the job
+	err = j.cmd.Start()
+	if err != nil {
+		return nil, fmt.Errorf("failed to start job's exec.Cmd: %w", err)
+	}
+	// Run go routine to handle the blocking call exec.Cmd.Wait() and update the running flag to indicate the job has complete
+	go func(runningJob *Job, logFile *os.File) {
+		runningJob.cmd.Wait()
+		runningJob.setRunning(false)
+		logFile.Close()
+	}(j, f)
+	return j, nil
+}
+
+// Stop request a job's termination using SIGTERM and deletes it's cgroup. If the SIGTERM is ignored, we send
+// a SIGKILL after STOP_GRACE_PERIOD.
+func (job *Job) Stop() error {
+	// Regardless of signalling errors, ensure we clean up the job's log file and cgroup
+	defer func() {
+		os.Remove(logPath(job.ID))
+		job.con.DeleteGroup(job.ID)
+	}()
+	// Request the process to terminate
+	if err := job.cmd.Process.Signal(syscall.SIGTERM); err != nil {
+		return fmt.Errorf("failed to send SIGTERM to job: %w", err)
+	}
+	// Poll the process with a signal 0 to check if it is running
+	running := true
+	killTime := time.Now().Add(STOP_GRACE_PERIOD)
+	for killTime.Unix() > time.Now().Unix() {
+		if err := job.cmd.Process.Signal(syscall.Signal(0)); err != nil {
+			running = false
+			break
+		}
+		time.Sleep(STOP_POLL_INTERVAL)
```

Comment:
Instead of polling, it might be more responsive to observe `runningJob.cmd.Wait()` completing.

Reply:
Good point, I now use a channel to signal when the job is done